federation/ipatuura/src/main/java/org/keycloak/ipatuura_user_spi/Ipatuura.java

```diff
@@ -0,0 +1,413 @@
+    }
+
+    public SCIMUser getUserByFirstName(String username) {
+        String attribute = "name.givenName";
+        return getUserByAttr(username, attribute);
+    }
+
+    public SCIMUser getUserByLastName(String username) {
+        String attribute = "name.familyName";
+        return getUserByAttr(username, attribute);
+    }
+
+    public Response deleteUser(String username) {
+        SCIMUser userobj = getUserByUsername(username);
+        SCIMUser.Resource user = userobj.getResources().get(0);
+
+        String userIdUrl = String.format("Users/%s", user.getId());
+
+        Response response;
+        try {
+            response = clientRequest(userIdUrl, "DELETE", null);
+        } catch (Exception e) {
+            logger.errorv("Error: {0}", e.getMessage());
+            throw new RuntimeException(e);
+        }
+
+        return response;
+    }
+
+    /*
+     * Keycloak UserRegistrationProvider addUser() method only provides username as input, here we provide mostly dummy values
+     * which will be replaced by actual user input via appropriate setter methods once in the returned UserModel
+     */
+    private SCIMUser.Resource setupUser(String username) {
+        SCIMUser.Resource user = new SCIMUser.Resource();
+        SCIMUser.Resource.Name name = new SCIMUser.Resource.Name();
+        SCIMUser.Resource.Email email = new SCIMUser.Resource.Email();
+        List<String> schemas = new ArrayList<String>();
+        List<SCIMUser.Resource.Email> emails = new ArrayList<SCIMUser.Resource.Email>();
+        List<SCIMUser.Resource.Group> groups = new ArrayList<SCIMUser.Resource.Group>();
+
+        schemas.add(SCHEMA_CORE_USER);
+        user.setSchemas(schemas);
+        user.setUserName(username);
+        user.setActive(true);
+        user.setGroups(groups);
+
+        name.setGivenName("dummyfirstname");
+        name.setMiddleName("");
+        name.setFamilyName("dummylastname");
+        user.setName(name);
+
+        email.setPrimary(true);
+        email.setType("work");
+        email.setValue("dummy@example.com");
+        emails.add(email);
+        user.setEmails(emails);
+
+        return user;
+    }
+
+    public Response createUser(String username) {
+        String usersUrl = "Users";
+
+        SCIMUser.Resource newUser = setupUser(username);
+
+        Response response;
+        try {
+            response = clientRequest(usersUrl, "POST", newUser);
+        } catch (Exception e) {
+            logger.errorv("Error: {0}", e.getMessage());
+            return null;
+        }
+
+        return response;
+    }
+
+    private void setUserAttr(SCIMUser.Resource user, String attr, String value) {
+        SCIMUser.Resource.Name name = user.getName();
+        SCIMUser.Resource.Email email = new SCIMUser.Resource.Email();
+        List<SCIMUser.Resource.Email> emails = new ArrayList<SCIMUser.Resource.Email>();
+
+        switch (attr) {
+            case "firstName":
```

Comment:
It is a nitpick but we are fetching root attribute names names from constants in `UserModel`. Mainly in user profile, where we usually want to see where those attributes are being used.

Reply:
You mean we should be using the constants here? I can fix that.

Reply:
Done!